lib/promise.mli

```diff
@@ -0,0 +1,13 @@
+type 'a promise
+val _Promise : 'a
+val _catch : 'a promise Js.t -> ('b -> 'a promise Js.t) -> 'a promise Js.t
+val _then :
+  'a promise Js.t ->
+  ?catch:('c -> 'b promise Js.t) ->
+  ('a -> 'b promise Js.t) -> 'b promise Js.t
+val resolve_value : 'a -> 'a promise Js.t
```

Comment:
As already mentioned in the PR's comment, this is where the unsoundness issue arises. If `resolve_value` is called at type `foo promise Js.t`, the result type will be `foo promise promise Js.t`. However, JS will produce a value whose representation corresponds to `foo promise Js.t`, i.e. it will silently "call" `join` to flatten the nested promise.

The user of this binding has to deliberately choose to call `resolve_promise` to encode this in OCaml. However, that is error-prone, and can't be done in polymorphic function.